src/subtyping.md

```diff
@@ -1,146 +1,143 @@
-makes it harder to give simple examples, which is a problem since subtyping,
-and especially variance, is already hard to understand properly. As in,
-even compiler writers mess it up all the time.
+In order to allow flexible usage of lifetimes
+while also preventing their misuse, Rust uses a combination of **Subtyping** and **Variance**.
 
-To keep things simple, this section will consider a small extension to the
-Rust language that adds a new and simpler subtyping relationship. After
-establishing concepts and issues under this simpler system,
-we will then relate it back to how subtyping actually occurs in Rust.
+Let's start with a example.
 
-So here's our simple extension, *Objective Rust*, featuring three new types:
-
-```rust
-trait Animal {
-    fn snuggle(&self);
-    fn eat(&mut self);
+```rust,ignore
+fn debug<T: std::fmt::Debug>(a: T, b: T) {
+    println!("a = {:?} b = {:?}", a, b);
 }
 
-trait Cat: Animal {
-    fn meow(&self);
+fn main() {
+    let hello: &'static str = "hello";
+    {
+        let world = String::from("world");
+        let world = &world; // 'b has a shorter lifetime than 'static
+        debug(hello, world);
+    }
 }
+```
 
-trait Dog: Animal {
-    fn bark(&self);
-}
+In a conservative implementation of lifetimes, since `a` and `b` have differeing lifetimes,
+we might see the following error:
+
+```text
+error[E0308]: mismatched types
+ --> src/main.rs:10:16
+   |
+10 |         debug(hello, world);
+   |                      ^
+   |                      |
+   |                      expected `&'static str`, found struct `&'b str`
 ```
 
-But unlike normal traits, we can use them as concrete and sized types, just like structs.
+This would be rather unfortunate. In this case,
+what we want is to accept any type that lives *at least as long* as `'b`.
+Let's try using subtyping with our lifetimes.
 
-Now, say we have a very simple function that takes an Animal, like this:
+## Subtyping
 
-<!-- ignore: simplified code -->
-```rust,ignore
-fn love(pet: Animal) {
-    pet.snuggle();
-}
-```
+Subtyping is the idea that one type can be used in place of another.
 
-By default, static types must match *exactly* for a program to compile. As such,
-this code won't compile:
+Let's define that `Sub` is a subtype of `Super` (we'll be using the notation `Sub: Super` throughout this chapter)
 
-<!-- ignore: simplified code -->
-```rust,ignore
-let mr_snuggles: Cat = ...;
-love(mr_snuggles);         // ERROR: expected Animal, found Cat
-```
+What this is suggesting to us is that the set of *requirements* that `Super` defines
+are completely satisfied by `Sub`. `Sub` may then have more requirements.
 
-Mr. Snuggles is a Cat, and Cats aren't *exactly* Animals, so we can't love him! 😿
+An example of simple subtyping that exists in the language are [supertraits](https://doc.rust-lang.org/stable/book/ch19-03-advanced-traits.html?highlight=supertraits#using-supertraits-to-require-one-traits-functionality-within-another-trait)
```

Comment:
```suggestion
An example of simple subtyping that exists in the language is [supertraits](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-supertraits-to-require-one-traits-functionality-within-another-trait):
```
It might also be useful to put the URL at the end using an anchor, as in `send-and-sync.md`.